File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {
	static int n, m;
	static Queue<Integer> q = new LinkedList<Integer>();
	static int[] check = new int[100001];
	static void find(int now, int next) {
		if(next<0||next>100000)return;
		if(check[next]>check[now]+1) {
			check[next]=check[now]+1;
			q.offer(next);
		}
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		n = sc.nextInt();
		m = sc.nextInt();
		q.offer(n);
		for (int i = 0; i < check.length; i++)check[i] = 10000000;
		check[n] = 0;
		while (!q.isEmpty()) {
			int now = q.poll();
			int[] next = { now - 1, now + 1, now * 2 };
			if (now == m)
				break;
			for (int i = 0; i < 3; i++) {
				find(now, next[i]);
			}
		}
		System.out.println(check[m]);
	}
}

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.cc
#include<iostream>
#include<algorithm>
#include<vector>
#include<queue>
#include<utility>
#include<cstring>
using namespace std;
int n, m, check[100005], cnt;
queue<int>q;
void find(int now, int next) {
	if (next < 0 || next>100000)return;
	if (check[next] > check[now] + 1) {
		check[next] = check[now] + 1;
		q.push(next);
	}
}
int main() {
	ios_base::sync_with_stdio(false), cin.tie(0), cout.tie(0);
	cin >> n >> m;
	q.push(n);
	for (int i = 0; i < 100005; i++)check[i] = 10000000;
	check[n] = 0;
	while (!q.empty()) {
		int now = q.front(), next[3] = { now - 1,now + 1,now * 2 };
		q.pop();
		if (now == m) {
			break;
		}
		for (int i = 0; i < 3; i++) {
			if (i && next[i] > 100000)break;
			find(now, next[i]);
		}
	}
	cout << check[m] << "\n";
	return 0;
}